### PR TITLE
Add pytest decorator plugin

### PR DESCRIPTION
**Pull request by MaximilianSoerenPollak**

### CODE REVIEW

**Comment by MaximilianSoerenPollak** [comment]
I want some feedback on this. 

I will implement some tests too

File: python_basics/score_pytest/attribute_plugin.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    if start_path == "":
+        start_path = __file__
+
+    git_root = Path(start_path).resolve()
+    esbonio_search = False
+    while not (git_root / ".git").exists():
+        git_root = git_root.parent
+        if git_root == Path("/"):
+            # fallback to cwd when building with python -m sphinx docs _build -T
+            if esbonio_search:
```
Comment (AlexanderLanin) [inline python_basics/score_pytest/attribute_plugin.py:23]: Please explain the logic of esbonio_search

File: python_basics/score_pytest/attribute_plugin.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    return git_root
+
+def add_test_properties(
+    *,
+    partially_verifies: list[str] | None = None,
+    fully_verifies: list[str] | None = None,
+    test_type: Literal["fault-injection", "interface-test", "requirements-based", "resource-usage"],
+    derivation_technique: Literal["requirements-analysis", "design-analysis", "boundary-values", "equivalence-classes", "fuzz-testing", "error-guessing", "explorative-testing"],
+    **extra_properties: Any,
+) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
```
Comment (AlexanderLanin) [inline python_basics/score_pytest/attribute_plugin.py:36]: I'm not sure this return type helps anyone. I would remove it.

File: python_basics/score_pytest/attribute_plugin.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        )    
+
+    # if not test_type:
+    #     raise ValueError("'test_type' is required and cannot be empty.")
+    #
+    # if not derivation_technique:
+    #     raise ValueError("'derivation_technique' is required and cannot be empty.")
+    #
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
```
Comment (AlexanderLanin) [inline python_basics/score_pytest/attribute_plugin.py:53]: Again, not sure about readability here. Maybe use Callable without parameters?

File: python_basics/score_pytest/attribute_plugin.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+def pytest_runtest_makereport(item: pytest.Item, call: pytest.CallInfo[None]) -> None:
+    """Attach file and line info to the report for use in junitxml output."""
+    if call.when != "call":
+        return
+
+    # Also support test_properties marker
+    marker = item.get_closest_marker("test_properties")
+    if marker and isinstance(marker.args[0], dict):
+        for k, v in marker.args[0].items():
+            item.user_properties.append((k, str(v)))
```
Comment (AlexanderLanin) [inline python_basics/score_pytest/attribute_plugin.py:83]: Feel free to comment this a little, as that's not something people encounter every day.

File: python_basics/score_pytest/attribute_plugin.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    recoursion_break = False
+    while not (git_root / ".git").exists():
+        git_root = git_root.parent
+        if git_root == Path("/"):
+            # fallback to cwd when building with python -m sphinx docs _build -T
+            if recoursion_break:
+                return None
+            git_root = Path.cwd().resolve()
+            recoursion_break = True
+    return git_root
```
Comment (MaximilianSoerenPollak) [inline python_basics/score_pytest/attribute_plugin.py:42]: Take the new one from the helper_lib

File: python_basics/score_pytest/attribute_plugin.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+
+@pytest.fixture(autouse=True)
+def add_file_and_line_attr(
+    record_xml_attribute: Callable[[str, str], None], request: pytest.FixtureRequest
+) -> None:
+    node = request.node
+    # print(node.__dir__())
+    # node.fspath gives the path to the test file
+    # node.file_path
```
Comment (MaximilianSoerenPollak) [inline python_basics/score_pytest/attribute_plugin.py:134]: Remove

File: python_basics/score_pytest/attribute_plugin.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    git_root = find_git_root(node.fspath)
+    file_path = Path(os.path.realpath(node.fspath)).relative_to(git_root)
+    # file_path = Path(node.fspath).relative_to(git_root)
+    # node.location is a tuple (filename, lineno, testname)
```
Comment (MaximilianSoerenPollak) [inline python_basics/score_pytest/attribute_plugin.py:138]: Remove

**Comment by MaximilianSoerenPollak** [comment]
README (attr_plugin_readme) is a first non proof read draft. 
I'm unsure what I should show here exactly. Any thoughts?

File: python_basics/score_pytest/ATTR_PLUGIN_README.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+This plugin lets you attach rich metadata to your test cases — like what requirements they verify, how they were derived, and what kind of test they are. That metadata ends up in your `junitxml` reports, making your testing output more meaningful, especially in traceability-driven projects.
+
+---
+
+## 🚀 How to Use
+
+Import the decorator:
+
+```python
+from your_plugin_module import add_test_properties
```
Comment (AlexanderLanin) [inline python_basics/score_pytest/ATTR_PLUGIN_README.md:14]: is it not?
```suggestion
from attribute_plugin import add_test_properties
```

File: python_basics/score_pytest/attribute_plugin.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+import pytest
+
+# Type aliases for better readability
+TestFunction = Callable[..., Any]
+Decorator = Callable[[TestFunction], TestFunction]
+
+
+
+def find_ws_root() -> Path | None:
```
Comment (AlexanderLanin) [inline python_basics/score_pytest/attribute_plugin.py:17]: this should use helper_lib?

Comment (MaximilianSoerenPollak) [reply python_basics/score_pytest/attribute_plugin.py:17]: Helper_lib is in docs_as_code. 

Though I could make one here too?

File: python_basics/score_pytest/ATTR_PLUGIN_README.md (hunk 8 above)
Comment (MaximilianSoerenPollak) [reply python_basics/score_pytest/ATTR_PLUGIN_README.md:14]: The README isn't 'checked' yet, it was just so that there is something rather than nothing. 
I'm sure there is more errors in it, will make sure to look it over.
